package/base-files/files/lib/functions/ipv4.sh

```diff
@@ -0,0 +1,166 @@
+uint_max=4294967295
+
+# Common error handler
+__die() {
+    echo "$1" >&2
+    exit 1
+}
+
+# check that $1 is only base 10 digits, and that it doesn't
+# exceed 2^32-1
+check_uint32() {
+    local n="$1"
+
+    [ -z "$n" -o -n "${n//[0-9]/}" ] && __die "Not a decimal integer ($n)"
+    [ "$n" -gt $uint_max ] && __die "Out of range ($n)"
+    [ "$((n + 0))" != "$n" ] && __die "Not normalized notation ($n)"
+
+    echo "$n"
+}
+
+# return a count of the number of bits set in $1
+bitcount() {
+    local c
+    c="$(check_uint32 "$1")" || exit 1
+
+    c=$((((c >> 1) & 0x55555555) + (c & 0x55555555)))
+    c=$((((c >> 2) & 0x33333333) + (c & 0x33333333)))
+    c=$((((c >> 4) & 0x0f0f0f0f) + (c & 0x0f0f0f0f)))
+    c=$((((c >> 8) & 0x00ff00ff) + (c & 0x00ff00ff)))
+    c=$((((c >> 16) & 0x0000ffff) + (c & 0x0000ffff)))
+
+    echo "$c"
+}
+
+# tedious but portable with busybox's limited shell
+# we check each octet to be in the range of 0..255,
+# and also make sure there's no extaneous characters.
+ip2int() {
+    local ip="$1" n ret=0
+
+    case "$ip" in
+    [0-9].*)
+	n="${ip:0:1}"
+	ip="${ip:2}"
+	;;
+    [1-9][0-9].*)
+	n="${ip:0:2}"
+	ip="${ip:3}"
+	;;
+    1[0-9][0-9].*|2[0-4][0-9].*|25[0-5].*)
+	n="${ip:0:3}"
+	ip="${ip:4}"
+	;;
+    *)
+	__die "Not a dotted quad ($1)"
+	;;
+    esac
+
+    ret=$((n << 24))
+
+    case "$ip" in
+    [0-9].*)
+	n="${ip:0:1}"
+	ip="${ip:2}"
+	;;
+    [1-9][0-9].*)
+	n="${ip:0:2}"
+	ip="${ip:3}"
+	;;
+    1[0-9][0-9].*|2[0-4][0-9].*|25[0-5].*)
+	n="${ip:0:3}"
+	ip="${ip:4}"
+	;;
+    *)
+	__die "Not a dotted quad ($1)"
+	;;
+    esac
+
+    ret=$((ret + (n << 16)))
+
+    case "$ip" in
+    [0-9].*)
+	n="${ip:0:1}"
+	ip="${ip:2}"
+	;;
+    [1-9][0-9].*)
+	n="${ip:0:2}"
+	ip="${ip:3}"
+	;;
+    1[0-9][0-9].*|2[0-4][0-9].*|25[0-5].*)
+	n="${ip:0:3}"
+	ip="${ip:4}"
+	;;
+    *)
+	__die "Not a dotted quad ($1)"
+	;;
+    esac
+
+    ret=$((ret + (n << 8)))
+
+    case "$ip" in
+    [0-9])
+	n="${ip:0:1}"
+	ip="${ip:1}"
+	;;
+    [1-9][0-9])
+	n="${ip:0:2}"
+	ip="${ip:2}"
+	;;
+    1[0-9][0-9]|2[0-4][0-9]|25[0-5])
+	n="${ip:0:3}"
+	ip="${ip:3}"
+	;;
+    *)
+	__die "Not a dotted quad ($1)"
+	;;
+    esac
+
+    ret=$((ret + n))
+
+    [ -n "$ip" ] && __die "Not a dotted quad ($1)"
+
+    echo "$ret"
+}
+
+# convert back from an integer to dotted-quad.
+int2ip() {
+    local n
+    n="$(check_uint32 "$1")" || exit 1
+
+    echo "$((n >> 24)).$(((n >> 16) & 255)).$(((n >> 8) & 255)).$((n & 255))"
```

Comment:
"ip" is a quad - "netmask" is an integer.
Not dramatic - but easy to improve now.
How about `ip_from_str`/`ip_to_str`?
`ip2str`/`str2ip`?

Reply:
`ip2int` was what was in the original awk code and no one had a problem with it, but I will rename it now so we can move forward.

Reply:
What kind of reasoning is that? It's an incremental process.
If no one had a problem with the old code, why this PR?
Anyway, I'm out.